### PR TITLE
refactor: Convert MultimediaEditFieldActivity to Kotlin

### PR DESCRIPTION
**Pull request by david-allison**
## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt [hunk 1]
```diff
@@ -16,549 +16,457 @@
...
-        mFieldController.onDone();
-
-        boolean bChangeToText = false;
-
-        if (mField.getType() == EFieldType.IMAGE) {
-            if (mField.getImagePath() == null) {
-                bChangeToText = true;
+    protected fun done() {
+        fieldController!!.onDone()
+        var bChangeToText = false
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt:203]: Maybe a cleanup to rename `bChangeToText` to remove the prefix? 
If not relevant to the repository's WIP kotlin code style, please ignore this.

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt [hunk 2]
```diff
@@ -16,549 +16,457 @@
...
-    private void saveAndExit() {
-        Intent resultData = new Intent();
-        resultData.putExtra(EXTRA_RESULT_FIELD, mField);
-        resultData.putExtra(EXTRA_RESULT_FIELD_INDEX, mFieldIndex);
-        setResult(RESULT_OK, resultData);
-        finishWithoutAnimation();
+    private fun saveAndExit() {
+        val resultData = Intent()
+        resultData.putExtra(EXTRA_RESULT_FIELD, mField)
+        resultData.putExtra(EXTRA_RESULT_FIELD_INDEX, mFieldIndex)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt:338]: Cleanup: `Scope function`

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt [hunk 3]
```diff
@@ -16,549 +16,457 @@
...
+        enableToolbar(mainView)
+        val intent = this.intent
+        mField = getFieldFromIntent(intent)
+        mNote = intent.getSerializableExtra(EXTRA_WHOLE_NOTE) as IMultimediaEditableNote?
+        mFieldIndex = intent.getIntExtra(EXTRA_FIELD_INDEX, 0)
         if (mField == null) {
-            UIUtils.showThemedToast(this, getString(R.string.multimedia_editor_failed), false);
-            finishCancel();
-            return;
+            showThemedToast(this, getString(R.string.multimedia_editor_failed), false)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt:87]: Keep `UIUtils.`

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt [hunk 4]
```diff
@@ -16,549 +16,457 @@
...
-
-            UIUtils.showThemedToast(this,
-                    getResources().getString(R.string.multimedia_editor_audio_permission_refused),
-                    true);
-
-            UIRecreationHandler.onRequiredPermissionDenied(mCurrentChangeRequest, this);
-
-        }
-        if (requestCode == REQUEST_CAMERA_PERMISSION && permissions.length == 1) {
+            showThemedToast(
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt:296]: Ditto

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt [hunk 5]
```diff
@@ -16,549 +16,457 @@
...
+                true
+            )
+            UIRecreationHandler.onRequiredPermissionDenied(mCurrentChangeRequest!!, this)
+        }
+        if (requestCode == REQUEST_CAMERA_PERMISSION && permissions.size == 1) {
             if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
-                UIUtils.showThemedToast(this,
-                        getResources().getString(R.string.multimedia_editor_camera_permission_refused),
-                        true);
+                showThemedToast(
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt:305]: Ditto

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt [hunk 6]
```diff
@@ -16,549 +16,457 @@
...
     }
 
-
-    private void cancelActivityWithAssertionFailure(String logMessage) {
-        Timber.e(logMessage);
-        UIUtils.showThemedToast(this, getString(R.string.mutimedia_editor_assertion_failed), false);
-        finishCancel();
+    private fun cancelActivityWithAssertionFailure(logMessage: String) {
+        Timber.e(logMessage)
+        showThemedToast(this, getString(R.string.mutimedia_editor_assertion_failed), false)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt:319]: Ditto

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt [hunk 7]
```diff
@@ -16,549 +16,457 @@
...
+        if (newUI.requiresPermissionCheck && performPermissionRequest(newUI.field)) {
+            newUI.markAsPermissionRequested()
+            return
         }
-
-        IControllerFactory controllerFactory = BasicControllerFactory.getInstance();
-
-        IFieldController fieldController = controllerFactory.createControllerForField(newUI.getField());
-
+        val controllerFactory = instance
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt:147]: Keep `BasicControllerFactory.`

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt [hunk 8]
```diff
@@ -16,549 +16,457 @@
...
-
-    private boolean performPermissionRequest(IField field) {
+    private fun performPermissionRequest(field: IField): Boolean {
         // Request permission to record if audio field
-        if (field instanceof AudioRecordingField && !Permissions.canRecordAudio(this)) {
-            Timber.d("Requesting Audio Permissions");
-            ActivityCompat.requestPermissions(this, new String[] {Manifest.permission.RECORD_AUDIO},
-                    REQUEST_AUDIO_PERMISSION);
-            return true;
+        if (field is AudioRecordingField && !canRecordAudio(this)) {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt:103]: Keep `Permissions.`

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt [hunk 9]
```diff
@@ -16,549 +16,457 @@
...
-
-    /** Sets various properties required for IFieldController to be in a valid state */
-    private void setupUIController(IFieldController fieldController, @Nullable Bundle savedInstanceState) {
-        fieldController.setField(mField);
-        fieldController.setFieldIndex(mFieldIndex);
-        fieldController.setNote(mNote);
-        fieldController.setEditingActivity(this);
-        fieldController.loadInstanceState(savedInstanceState);
+        if (field is ImageField && manifestContainsPermission(this) &&
+            !canUseCamera(this)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt:114]: Keep `Permissions.`

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt [hunk 10]
```diff
@@ -16,549 +16,457 @@
...
-    }
-
-    /** Sets various properties required for IFieldController to be in a valid state */
-    private void setupUIController(IFieldController fieldController, @Nullable Bundle savedInstanceState) {
-        fieldController.setField(mField);
-        fieldController.setFieldIndex(mFieldIndex);
-        fieldController.setNote(mNote);
-        fieldController.setEditingActivity(this);
-        fieldController.loadInstanceState(savedInstanceState);
+        if (field is ImageField && manifestContainsPermission(this) &&
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt:113]: Keep `CheckCameraPermission.`

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt [hunk 11]
```diff
@@ -16,549 +16,457 @@
...
 
-    private void recreateEditingUi(ChangeUIRequest newUI) {
-        this.recreateEditingUi(newUI, null);
+    /** Sets various properties required for IFieldController to be in a valid state  */
+    private fun setupUIController(fieldController: IFieldController, savedInstanceState: Bundle?) {
+        fieldController.setField(mField!!)
+        fieldController.setFieldIndex(mFieldIndex)
+        fieldController.setNote(mNote)
+        fieldController.setEditingActivity(this)
+        fieldController.loadInstanceState(savedInstanceState)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.kt:132]: Cleanup: `scope function`

**Review by BrayanDSO — CHANGES_REQUESTED** [review]
I'd like this one in soon, so here's some help.

**Comment by david-allison** [comment]
I'll get this one fixed up tomorrow, thanks for the review

**Review by mikehardy — APPROVED** [review]
I used this as a test and I think I would have caught 2/3rds, maybe 3/4s of what @BrayanDSO saw. I did not see anything else. nice work

**Comment by mikehardy** [comment]
tested locally and everything worked - the unit test failure is likely a flake, I'll note it

**Pull request merged**